app/controllers/pensions/paymentsIntoOverseasPensions/QOPSReferenceController.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.pensions.paymentsIntoOverseasPensions
+
+import config.{AppConfig, ErrorHandler}
+import controllers.predicates.AuthorisedAction
+import forms.QOPSReferenceNumberForm
+import models.User
+import play.api.data.Form
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import services.PensionSessionService
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+import utils.Clock
+import views.html.pensions.paymentsIntoOverseasPensions.QOPSReferenceView
+import controllers.pensions.routes.PensionsSummaryController
+import controllers.pensions.paymentsIntoOverseasPensions.routes.QOPSReferenceController
+import models.mongo.PensionsCYAModel
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class QOPSReferenceController @Inject()(authAction: AuthorisedAction,
+                                        qopsReferenceView: QOPSReferenceView,
+                                        pensionSessionService: PensionSessionService,
+                                        errorHandler: ErrorHandler
+                                       )(implicit val mcc: MessagesControllerComponents,
+                                         appConfig: AppConfig,
+                                         clock: Clock,
+                                         ec: ExecutionContext) extends FrontendController(mcc) with I18nSupport {
+  def referenceForm(user: User): Form[String] = QOPSReferenceNumberForm.qopsReferenceNumberForm(
+    incorrectFormatMsg = "overseasPension.QOPSReference.error.incorrectFormat"
+  )
+
+  def show(taxYear: Int): Action[AnyContent] = authAction.async { implicit request =>
```

Comment:
I mentioned using `map` during the group review.

I meant something like the following:

```

  def show(taxYear: Int): Action[AnyContent] = authAction.async { implicit request =>
    pensionSessionService.getPensionSessionData(taxYear, request.user).flatMap {
      case Left(_) => Future.successful(errorHandler.handleError(INTERNAL_SERVER_ERROR))
      case Right(optPensionUserData) =>
        Future.successful(
          optPensionUserData
            .map(pensionUserData => Ok(qopsReferenceView(referenceForm(request.user, pensionUserData), taxYear)))
            .getOrElse(Redirect(PensionsSummaryController.show(taxYear))))
    }
  }

  private def referenceForm(user: User, pensionUserData: PensionsUserData)(implicit request: AuthorisationRequest[AnyContent]): Form[String] =
    pensionUserData.pensions.paymentsIntoOverseasPensions.qualifyingOverseasPensionSchemeReferenceNumber
      .map(value => referenceForm(user).fill(removePrefix(value)))
      .getOrElse(referenceForm(user))
```

Or, if inlined:

```
 def show(taxYear: Int): Action[AnyContent] = authAction.async { implicit request =>
    pensionSessionService.getPensionSessionData(taxYear, request.user).flatMap {
      case Left(_) => Future.successful(errorHandler.handleError(INTERNAL_SERVER_ERROR))
      case Right(optPensionUserData) =>
        Future.successful(
          optPensionUserData
            .map(pensionUserData =>
              Ok(
                qopsReferenceView(
                  pensionUserData.pensions.paymentsIntoOverseasPensions.qualifyingOverseasPensionSchemeReferenceNumber
                    .map(value => referenceForm(request.user).fill(removePrefix(value)))
                    .getOrElse(referenceForm(request.user)), taxYear)))
            .getOrElse(Redirect(PensionsSummaryController.show(taxYear))))
    }
  
```

Reply:
Actually, we can remove the two `Future.successful` calls; we were stripping them out with the `flatMap`:

```
 def show(taxYear: Int): Action[AnyContent] = authAction.async { implicit request =>
    pensionSessionService.getPensionSessionData(taxYear, request.user).map {
      case Left(_) => errorHandler.handleError(INTERNAL_SERVER_ERROR)
      case Right(optPensionUserData) =>
        optPensionUserData
          .map(pensionUserData =>
            Ok(
              qopsReferenceView(
                pensionUserData.pensions.paymentsIntoOverseasPensions.qualifyingOverseasPensionSchemeReferenceNumber
                  .map(value => referenceForm(request.user).fill(removePrefix(value)))
                  .getOrElse(referenceForm(request.user)), taxYear)))
          .getOrElse(Redirect(PensionsSummaryController.show(taxYear)))
    }
  }
```

Reply:
For what it's worth, my preference is for:

```
 def show(taxYear: Int): Action[AnyContent] = authAction.async { implicit request =>
    pensionSessionService.getPensionSessionData(taxYear, request.user).map {
      case Left(_) => errorHandler.handleError(INTERNAL_SERVER_ERROR)
      case Right(optPensionUserData) =>
        optPensionUserData
          .map(pensionUserData => Ok(qopsReferenceView(referenceForm(request.user, pensionUserData), taxYear)))
          .getOrElse(Redirect(PensionsSummaryController.show(taxYear)))
    }
  }

  private def referenceForm(user: User, pensionUserData: PensionsUserData)(implicit request: AuthorisationRequest[AnyContent]): Form[String] =
    pensionUserData.pensions.paymentsIntoOverseasPensions.qualifyingOverseasPensionSchemeReferenceNumber
      .map(value => referenceForm(user).fill(removePrefix(value)))
      .getOrElse(referenceForm(user))
```
Why?

Besides the reduction of duplication, the 'some' and 'none' cases are each one line, so symmetrical.

Reply:
@erasmos-hmrc so do you want the code amended as part of this CR? Or shall we have a Spike to discuss this?

Reply:
It's up Salman & Hassan.

No need of a spike :) Only a suggestion.

Reply:
Preferred the last suggested implementation and used it. Thanks @erasmos-hmrc

Reply:
My pleasure.

Confirmed.